In [ ]:
C:\Users\asli_\PYTHON_PROJECTS

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
sonar_data = pd.read_csv('Copy of sonar data.csv', header=None)

In [3]:
sonar_data.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      51      52      53      54      55      56      57  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

       58      59  60  
0  0.0090  0.0032   R  
1  0.0052  0.0044   R  
2  0.0095  0.0078   R  
3  0.0040  0.0117   R  
4  0.0107  0.0094   R  

[5 rows x 61 columns]

Bu veri, **sonar sinyalleri** kullanılarak **kayaları (R)** ve **mayınları (M)** ayırt etmeyi amaçlayan bir veri setidir. Her satır, bir sonar sinyalinin çeşitli frekanslardaki ölçümlerini temsil eder ve son sütun, sinyalin bir kayadan mı yoksa mayından mı yansıdığını gösterir.

### Verideki Detaylar:
- **0-59 arası sütunlar**: Sinyalin farklı frekanslardaki yansımalarını temsil eden sayısal değerlerdir. Yani, her sütun bir frekansa ait sinyal gücünü gösterir.
- **60. sütun**: Son sütun, sinyalin yansıdığı nesne türünü belirtir:
  - `R`: Kaya (Rock)
  - `M`: Mayın (Mine)

### Gözlemler:
- **Değerlerin aralığı**: 0 ile 1 arasında değişen küçük sayılar. Bu, sinyal genliklerinin normalleştirildiğini gösteriyor.
- **Etiketler**: Verinin sınıflandırma problemi için hazır olduğu görülüyor çünkü her satırın sonunda bir etiket (R veya M) var.

Sonuç olarak, bu veri, sonar sinyallerinin hangi tür yüzeyden yansıdığını tahmin etmek için kullanılıyor ve makine öğrenmesi modelleriyle çalışmak için uygun.

In [4]:
sonar_data.shape

(208, 61)

In [5]:
sonar_data.describe()

0           1           2           3           4           5   \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.029164    0.038437    0.043832    0.053892    0.075202    0.104570   
std      0.022991    0.032960    0.038428    0.046528    0.055552    0.059105   
min      0.001500    0.000600    0.001500    0.005800    0.006700    0.010200   
25%      0.013350    0.016450    0.018950    0.024375    0.038050    0.067025   
50%      0.022800    0.030800    0.034300    0.044050    0.062500    0.092150   
75%      0.035550    0.047950    0.057950    0.064500    0.100275    0.134125   
max      0.137100    0.233900    0.305900    0.426400    0.401000    0.382300   

               6           7           8           9   ...          50  \
count  208.000000  208.000000  208.000000  208.000000  ...  208.000000   
mean     0.121747    0.134799    0.178003    0.208259  ...    0.016069   
std      0.061788    0.085152    0.118387    0.134416  ...    0.012008   
min      0.003300    0.005500    0.007500    0.011300  ...    0.000000   
25%      0.080900    0.080425    0.097025    0.111275  ...    0.008425   
50%      0.106950    0.112100    0.152250    0.182400  ...    0.013900   
75%      0.154000    0.169600    0.233425    0.268700  ...    0.020825   
max      0.372900    0.459000    0.682800    0.710600  ...    0.100400   

               51          52          53          54          55          56  \
count  208.000000  208.000000  208.000000  208.000000  208.000000  208.000000   
mean     0.013420    0.010709    0.010941    0.009290    0.008222    0.007820   
std      0.009634    0.007060    0.007301    0.007088    0.005736    0.005785   
min      0.000800    0.000500    0.001000    0.000600    0.000400    0.000300   
25%      0.007275    0.005075    0.005375    0.004150    0.004400    0.003700   
50%      0.011400    0.009550    0.009300    0.007500    0.006850    0.005950   
75%      0.016725    0.014900    0.014500    0.012100    0.010575    0.010425   
max      0.070900    0.039000    0.035200    0.044700    0.039400    0.035500   

               57          58          59  
count  208.000000  208.000000  208.000000  
mean     0.007949    0.007941    0.006507  
std      0.006470    0.006181    0.005031  
min      0.000300    0.000100    0.000600  
25%      0.003600    0.003675    0.003100  
50%      0.005800    0.006400    0.005300  
75%      0.010350    0.010325    0.008525  
max      0.044000    0.036400    0.043900  

[8 rows x 60 columns]

In [6]:
sonar_data[60].value_counts()

60
M    111
R     97
Name: count, dtype: int64

Bu satırda, **SONAR veri setinin 60. sütunundaki etiketlerin (M ve R)** frekansları sayılmış ve sonuç aşağıdaki gibi elde edilmiştir:

- **M (Mines)**: 111 kez gözlemlenmiş.  
- **R (Rocks)**: 97 kez gözlemlenmiş.

In [7]:
sonar_data.groupby(60).mean()

0         1         2         3         4         5         6   \
60                                                                         
M   0.034989  0.045544  0.050720  0.064768  0.086715  0.111864  0.128359   
R   0.022498  0.030303  0.035951  0.041447  0.062028  0.096224  0.114180   

          7         8         9   ...        50        51        52        53  \
60                                ...                                           
M   0.149832  0.213492  0.251022  ...  0.019352  0.016014  0.011643  0.012185   
R   0.117596  0.137392  0.159325  ...  0.012311  0.010453  0.009640  0.009518   

          54        55        56        57        58        59  
60                                                              
M   0.009923  0.008914  0.007825  0.009060  0.008695  0.006930  
R   0.008567  0.007430  0.007814  0.006677  0.007078  0.006024  

[2 rows x 60 columns]

In [8]:
X = sonar_data.drop(columns=60, axis=1)
Y = sonar_data[60]

In [9]:
print(X)
print(Y)

         0       1       2       3       4       5       6       7       8   \
0    0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1    0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2    0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3    0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4    0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      50      51      52      53   

--

### **Adımların Detaylı Açıklaması:**

1. **Verileri Hazırlama:**
   - **X (Girdiler)**: 0-59 sütunları, yani ses sinyalinin frekans genlikleri. Model bu verilerden öğrenir.
   - **Y (Çıktı/Etiket)**: 60. sütun, yani gözlemin **Mine (M)** veya **Rock (R)** olduğunu belirten sınıf etiketi.

2. **Modeli Eğitme:**
   - Model, **X'teki sinyal verilerini** kullanarak **Y'nin sınıfını** tahmin etmeyi öğrenir.
   - Bu süreçte model, **girdiler (X)** ile **etiketler (Y)** arasındaki ilişkileri keşfetmeye çalışır.

3. **Tahmin Yapma:**
   - Eğitim tamamlandıktan sonra model, yeni verilere (yeni X değerlerine) bakarak bir tahmin yapar. Bu tahmin, gözlemin **Mine** mı yoksa **Rock** mu olduğunu söyler.

4. **Tahmin Sonucunu Değerlendirme:**
   - Modelin yaptığı tahminler, gerçek etiketlerle (**Y'nin gerçek değerleriyle**) karşılaştırılır.
   - **Hedef:** Modelin tahminlerinin doğruluğunu ölçmek ve hataları azaltmak.

---

### **Bir Örnekle Açıklama:**

- **Eğitim Verisi (X):**  
  - Girdiler: `[0.0200, 0.0371, 0.0428, ..., 0.0032]` → Ses sinyali verileri.
  
- **Etiket (Y):**  
  - Gerçek sınıf: `'R'` (Rock).

- **Modelin Tahmini:**  
  - Model bu girdileri kullanarak `'R'` (Rock) olduğunu tahmin eder.

- **Değerlendirme:**  
  - Modelin tahmini gerçek değerle (Y) karşılaştırılır. Eğer doğruysa, başarı; yanlışsa hata olarak kaydedilir.

---

### **Sonuç:**
- **Amaç:** Modeli eğiterek, sinyal verilerine (X) bakarak doğru sınıfı (Y) tahmin edebilmesini sağlamaktır. Tahmin edilen Y ile gerçek Y'yi karşılaştırarak modelin performansını ölçeriz.

Bu döngüyü iyileştirmek, modelin doğruluğunu artırır! 🚀

In [10]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.1, stratify=Y, random_state=1)

In [11]:
print(X.shape, X_train.shape, X_test.shape)

(208, 60) (187, 60) (21, 60)


In [12]:
print(X_train)
print(Y_train)

         0       1       2       3       4       5       6       7       8   \
115  0.0414  0.0436  0.0447  0.0844  0.0419  0.1215  0.2002  0.1516  0.0818   
38   0.0123  0.0022  0.0196  0.0206  0.0180  0.0492  0.0033  0.0398  0.0791   
56   0.0152  0.0102  0.0113  0.0263  0.0097  0.0391  0.0857  0.0915  0.0949   
123  0.0270  0.0163  0.0341  0.0247  0.0822  0.1256  0.1323  0.1584  0.2017   
18   0.0270  0.0092  0.0145  0.0278  0.0412  0.0757  0.1026  0.1138  0.0794   
..      ...     ...     ...     ...     ...     ...     ...     ...     ...   
140  0.0412  0.1135  0.0518  0.0232  0.0646  0.1124  0.1787  0.2407  0.2682   
5    0.0286  0.0453  0.0277  0.0174  0.0384  0.0990  0.1201  0.1833  0.2105   
154  0.0117  0.0069  0.0279  0.0583  0.0915  0.1267  0.1577  0.1927  0.2361   
131  0.1150  0.1163  0.0866  0.0358  0.0232  0.1267  0.2417  0.2661  0.4346   
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   

         9   ...      50      51      52      53   

In [13]:
model = LogisticRegression()

In [14]:
model.fit(X_train, Y_train)

LogisticRegression()

In [15]:
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train) 

In [16]:
print('Accuracy on training data : ', training_data_accuracy)

Accuracy on training data :  0.8342245989304813


#### **Eğitim (Training):**
```python
model = LogisticRegression()
model.fit(X_train, Y_train)
```
- Bu adımda, **Logistik Regresyon modelini** oluşturuyorsunuz ve modelin eğitimini **X_train** (özellikler) ve **Y_train** (etiketler) verisiyle yapıyorsunuz.
- **model.fit()** fonksiyonu, modelin **özellikler ile etiketler arasındaki ilişkiyi öğrenmesini** sağlar. Bu aşama, modelin sınıfları (Mines ve Rocks) daha iyi ayırt edebilmesi için bir model oluşturur.

#### **Modelin Performansını Değerlendirme:**
```python
model.score(X_test, Y_test)
```
- **score()** fonksiyonu, modelin **test verisi** üzerinde ne kadar başarılı olduğunu gösterir. Bu, modelin genel **doğruluk oranını** ölçer.
  - **X_test** ve **Y_test**, **test veri kümesi** üzerindeki özellikler ve gerçek etiketlerdir.
  - Test doğruluğu: **0.8333 (yaklaşık %83)**. Bu, modelin test veri kümesindeki örnekleri doğru tahmin etme oranıdır.

```python
model.score(X_train, Y_train)
```
- Eğitim verisi üzerindeki doğruluk: **0.8072 (yaklaşık %81)**. Bu, modelin **eğitim verisi** üzerindeki doğruluğudur.
  
- Bu sonuçlar şunları gösteriyor:
  - Eğitim verisi üzerinde doğruluk oranı biraz daha düşükken, test verisi üzerinde doğruluk oranı **%83** civarında. Bu, modelin **genelleme** yeteneğinin iyi olduğunu ve **overfitting (aşırı öğrenme)** yapmadığını gösteriyor. Yani, model **eğitim verisinde iyi performans gösterdiği kadar test verisinde de başarılı**.

---

### **Sonuçların Yorumlanması:**

- **Modelin Test Performansı (0.8333)**: Test verisi üzerindeki doğruluk oranı oldukça yüksek, yani model **yeni ve görünmeyen verilerle iyi sonuçlar** veriyor. Bu, modelin **genelleme kapasitesinin iyi olduğunu** gösterir.
  
- **Modelin Eğitim Performansı (0.8072)**: Eğitim verisi üzerindeki doğruluk oranı biraz daha düşük. Bu, modelin eğitim verisini öğrenip, daha fazla **özelliklerin etkisini** anlamaya başladığını, ancak aynı zamanda test verisinde **daha iyi sonuçlar** verdiğini gösterir.

### **Logistik Regresyonun Seçilmesinin Nedeni:**
- **Verinin doğrusal doğası**: Sonar verisi, **ikili sınıflandırma** ve **doğrusal sınırlarla** iyi ayırt edilebilecek veriler içeriyorsa, logistik regresyon iyi sonuçlar verebilir.
---


In [17]:
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [18]:
print('Accuracy on test data : ', test_data_accuracy)

Accuracy on test data :  0.7619047619047619


In [19]:
input_data = (0.0307,0.0523,0.0653,0.0521,0.0611,0.0577,0.0665,0.0664,0.1460,0.2792,0.3877,0.4992,0.4981,0.4972,0.5607,0.7339,0.8230,0.9173,0.9975,0.9911,0.8240,0.6498,0.5980,0.4862,0.3150,0.1543,0.0989,0.0284,0.1008,0.2636,0.2694,0.2930,0.2925,0.3998,0.3660,0.3172,0.4609,0.4374,0.1820,0.3376,0.6202,0.4448,0.1863,0.1420,0.0589,0.0576,0.0672,0.0269,0.0245,0.0190,0.0063,0.0321,0.0189,0.0137,0.0277,0.0152,0.0052,0.0121,0.0124,0.0055)

# changing the input_data to a numpy array
input_data_as_numpy_array = np.asarray(input_data)

# reshape the np array as we are predicting for one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

prediction = model.predict(input_data_reshaped)
print(prediction)

if (prediction[0]=='R'):
  print('The object is a Rock')
else:
  print('The object is a mine')

['M']
The object is a mine


### Durum:
- **60 tane sütun (özellik)** var.
- Eğer 45 tanesi **'R' (Rock)** sınıfına aitse, **kalan 15 tanesi ise 'M' (Mine)** sınıfına aitse, **bu durum şu şekilde değerlendirilir**:

### Modelin Tahmin Yöntemi:
1. **Logistik regresyon** modeli her bir sütun için olasılık değerleri hesaplar. Bu, her sütunun **'R'** veya **'M'** sınıfına ait olma olasılıklarıdır.
2. Modelin genel **tahmininin ne olacağı**, her sütun için hesaplanan **olay olasılıklarının bir araya gelmesiyle belirlenir**. Modelin verdiği olasılık değeri **0 ile 1 arasında bir sayıdır**. 
   - Eğer olasılık **0.5'in üstünde** ise, model o örneği **'M' (Mine)** sınıfına atar.
   - Eğer olasılık **0.5'in altında** ise, model o örneği **'R' (Rock)** sınıfına atar.

### Durumun Gerçek Yorumlanması:
- Eğer **60 sütundan 45'i 'R' (Rock)** sınıfına aitse, bu durumda modelin büyük olasılıkla **'R' (Rock)** sınıfına ait olduğunu tahmin etmesi beklenir.
- Çünkü çoğunluk **'R' (Rock)** sınıfından oluşuyor ve model bu durumu daha olası olarak değerlendirir. Modelin tahmin ettiği olasılık, bu durumda **'R' (Rock)** sınıfına daha yakın olacaktır.


 Burada aslında **"tahminleme"** yapmıyoruz, çünkü elimizdeki veriler zaten etiketlenmiş ve her bir örneğin **"Rock" (R) veya "Mine" (M)** olduğu belli. Ancak, yapılan işlem şudur:

### 1. **Model Eğitimi (Training)**:
- Elimizde **etiketlenmiş veriler** var. Bu verilerde her bir satırda 60 özellik (sinyal verileri) ve 1 etiket (R veya M) bulunuyor.
- **Logistik regresyon** modeli bu verileri kullanarak eğitildi. Model, sinyal verilerinin (özelliklerin) **"Rock" (R)** veya **"Mine" (M)** sınıfına ait olup olmadığını öğrenmeye çalıştı.

### 2. **Tahmin (Prediction)**:
- Elimizde bir **yeni örnek veri** (yani **input_data**) olduğunda, model bu veriyi kullanarak bir tahminde bulunur. Model, bu örneğin **"Rock" (R)** mi yoksa **"Mine" (M)** mi olduğunu tahmin eder.
  
- Yani bu durumda model, daha önce **eğitilen verilerden öğrendiği ilişkiler** doğrultusunda bu yeni veriyi **'R' (Rock)** ya da **'M' (Mine)** sınıfına yerleştirir.


### Özet:
- **Tahminleme** yaptık çünkü model yeni veriyi kullanarak **'R' (Rock)** ya da **'M' (Mine)** sınıflandırması yaptı.
- Bu sınıflandırma aslında **doğru etiketleme** yapıldığının kontrolüydü, çünkü zaten verilerin etiketleri biliniyordu (sonuçları doğrulamak için).
pacaktır.

### Örneğin Durum:
- Diyelim ki eğitim verilerinizde **0.0523** değeri ile etiket **"M"** olarak belirlenmiş.
- Eğitilen model, **0.0523** değerini gördüğünde, **"M" (Mine)** etiketini tahmin eder.
- Bu etiketlemenin ardından **modelin doğruluğunu test etmek için** tahmin edilen etiket ile gerçek etiket (sınıf) karşılaştırılır. Eğer model doğru tahminde bulunmuşsa, doğruluk oranı yükseruluğunu ölçer**.
uğunu kontrol ettik.